### PR TITLE
Add Triangular distribution

### PR DESCRIPTION
**Pull request by DenisKolodin**
This PR contains `Triangular` [distribution](https://en.wikipedia.org/wiki/Triangular_distribution) implementation. Check it, please.

### CODE REVIEW

File: src/distribution/mod.rs [hunk 1]
```diff
@@ -101,6 +101,7 @@ mod exponential;
 mod gamma;
 mod gaussian;
 mod uniform;
+mod triangular;
```
Comment (IvanUkhov) [inline src/distribution/mod.rs]: Let’s keep this list with `mod` and the one below with `pub use` sorted in alphabetical order.

File: src/distribution/triangular.rs [hunk 2]
```diff
@@ -0,0 +1,241 @@
+use distribution;
+use source::Source;
+
+/// An triangualr distribution.
```
Comment (IvanUkhov) [inline src/distribution/triangular.rs]: There are a couple of typos here.

File: src/distribution/triangular.rs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+#[derive(Clone, Copy)]
+pub struct Triangular {
+    a: f64,
+    b: f64,
+    c: f64,
+}
+
+impl Triangular {
+    pub fn new(a: f64, b: f64, c: f64) -> Self {
```
Comment (IvanUkhov) [inline src/distribution/triangular.rs]: Let’s add a description here, mentioning also that it panics if the arguments are invalid. Have a look at `new` of any other distribution.

Comment (IvanUkhov) [inline src/distribution/triangular.rs]: I also think that such simple functions should have `#[inline]`.

File: src/distribution/triangular.rs [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        Triangular {
+            a: a,
+            b: b,
+            c: c,
+        }
+    }
+}
+
+impl distribution::Continuous for Triangular {
+    #[inline]
```
Comment (IvanUkhov) [inline src/distribution/triangular.rs]: I wouldn’t inline here. The function seems rather big.

File: src/distribution/triangular.rs [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+            0.0
+        } else if self.b <= x {
+            1.0
+        } else {
+            let diff = self.b - self.a;
+            match x.partial_cmp(&self.c) {
+                Some(Less) | Some(Equal) =>
+                    (x - self.a).powi(2) / diff / (self.c - self.a),
+                Some(Greater) =>
+                    1.0 - (self.b - x).powi(2) / diff / (self.b - self.c),
```
Comment (IvanUkhov) [inline src/distribution/triangular.rs]: According to the [style guide](https://github.com/rust-lang/rust/blob/master/src/doc/style/style/whitespace.md), one can use up to 99 characters per line. It should probably fit on one line.

File: src/distribution/triangular.rs [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+                    f64::NAN,
+            }
+        }
+    }
+}
+
+impl distribution::Distribution for Triangular {
+    type Value = f64;
+
+    #[inline]
```
Comment (IvanUkhov) [inline src/distribution/triangular.rs]: I wound’t inline here.

File: src/distribution/triangular.rs [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        use std::f64;
+        use std::cmp::Ordering::*;
+        let &Triangular { a, b, c } = self;
+        if x < a || b < x {
+            0.0
+        } else {
+            let factor = 2.0 / (b - a);
+            match x.partial_cmp(&c) {
+                Some(Less) =>
+                    factor * (x - a) / (c - a),
```
Comment (IvanUkhov) [inline src/distribution/triangular.rs]: I’d put it on the same line with `=>`.

File: src/distribution/triangular.rs [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+impl distribution::Entropy for Triangular {
+    #[inline]
+    fn entropy(&self) -> f64 {
+        0.5 + ((self.b - self.a) / 2.0).ln()
+    }
+}
+
+impl distribution::Inverse for Triangular {
+    #[inline]
```
Comment (IvanUkhov) [inline src/distribution/triangular.rs]: I wouldn’t inline here.

File: src/distribution/triangular.rs [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+impl distribution::Sample for Triangular {
+    #[inline]
+    fn sample<S>(&self, source: &mut S) -> f64 where S: Source {
+        use distribution::Inverse;
+        self.inverse(source.read::<f64>())
+    }
+}
+
+impl distribution::Skewness for Triangular {
+    #[inline]
```
Comment (IvanUkhov) [inline src/distribution/triangular.rs]: Perhaps I wouldn’t inline here either.

File: src/distribution/triangular.rs [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+#[cfg(test)]
+mod tests {
+    use assert;
+    use prelude::*;
+
+    macro_rules! new(
+        ($a:expr, $b:expr, $c:expr) => (Triangular::new($a, $b, $c));
+    );
+
+
```
Comment (IvanUkhov) [inline src/distribution/triangular.rs]: There are two empty lines here.

File: src/prelude.rs [hunk 11]
```diff
@@ -23,6 +23,7 @@ pub use distribution::Exponential;
 pub use distribution::Gamma;
 pub use distribution::Gaussian;
 pub use distribution::Uniform;
+pub use distribution::Triangular;
```
Comment (IvanUkhov) [inline src/prelude.rs]: Could you please sort this list?

**Comment by IvanUkhov** [comment]
Thank you! Except for those minor issues, it looks really solid!

**Comment by DenisKolodin** [comment]
Done! Also I add getters for `a`, `b`, `c` fields. Most of inlines I've removed. Check it If I removed odds, please.

**Comment by DenisKolodin** [comment]
According whitespace guide shall I remove spaces between elements here?
```rust
let p = vec![0.0, 0.03125, 0.125, 0.28125, 0.5, 0.71875, 0.875, 0.96875, 1.0];
let x = vec![1.0,     1.5,   2.0,     2.5, 3.0,     3.5,   4.0,     4.5, 5.0];
```

**Comment by IvanUkhov** [comment]
Regarding those spaces, I’d remove them. I break this rule only when doing so substantially improves legibility (like when you have a massive array of data). I don’t think it’s needed here.

**Comment by DenisKolodin** [comment]
I've add `nonnan` macro and changes for `Triangular` distribution as separated commit.
Also I've rebased previous commit with distribution implementation to remove spaces.

I think we don't have to squash both, because there is different meaning of changes.

**Comment by IvanUkhov** [comment]
Thank you!

**Pull request merged**